### ISSUE TITLE
Red squiggles under std::begin and std::end that shouldn't be there

### ISSUE THREAD

**Issue by papb** [microsoft/vscode-cpptools#1736]
* Operating System: Windows 7
* VS Code version: Code 1.21.1 (79b44aa704ce542d8ca4a3cc44cfca566e7720f1, 2018-03-14T14:46:47.128Z)
* C/C++ extension version: 0.15.0
* How to reproduce: create a file `index.cpp` with the following:

```c++
#include <iostream>
#include <algorithm>
int main(int argc, char **argv) {
    int v[3];
    std::cin >> v[0];
    std::cin >> v[1];
    std::cin >> v[2];
    std::sort(std::begin(v), std::end(v));
    return EXIT_SUCCESS;
}
```

* The problem:

![https://i.imgur.com/Gc1YbTR.png](https://i.imgur.com/Gc1YbTR.png)

The problem is that there is no problem with that code. There shouldn't be any red squiggles there.

* `c_cpp_properties.json`:

```json
{
    "configurations": [
        {
            "name": "Win32",
            "includePath": [
                "C:/Cygwin/lib/gcc/x86_64-pc-cygwin/6.4.0/include/c++/x86_64-pc-cygwin/",
                "C:/Cygwin/lib/gcc/x86_64-pc-cygwin/6.4.0/include/c++",
                "C:/Cygwin/lib/gcc/x86_64-pc-cygwin/6.4.0/include/",
                "C:/Cygwin/usr/include",
                "C:/Cygwin/usr/local/include",
                "${workspaceRoot}"
            ],
            "defines": [
                "_DEBUG",
                "UNICODE",
                "_UNICODE",
                "__gnuc__=6",
                "__cdecl=__attribute__((__cdecl__))"
            ],
            "intelliSenseMode": "clang-x64",
            "browse": {
                "path": [
                    "C:/Cygwin/lib/gcc/x86_64-pc-cygwin/6.4.0/include/c++/x86_64-pc-cygwin/",
                    "C:/Cygwin/lib/gcc/x86_64-pc-cygwin/6.4.0/include/c++",
                    "C:/Cygwin/lib/gcc/x86_64-pc-cygwin/6.4.0/include/",
                    "C:/Cygwin/usr/include",
                    "C:/Cygwin/usr/local/include",
                    "${workspaceRoot}"
                ],
                "limitSymbolsToIncludedHeaders": true,
                "databaseFilename": ""
            }
        }
    ],
    "version": 3
}
```

**Comment by sean-mcmanus** [comment]
Can you try https://github.com/Microsoft/vscode-cpptools/releases/tag/v0.16.0-insiders2 and set the compilerPath to your gcc? I was getting that error, but it went away after I set the compilerPath and cppStandard correctly (although I used clang). If your compiler doesn't support c++17 you may need to use c++14 instead (looks like we have some bug on Windows that isn't falling back properly, which we'll look into fixing before our final release).

**Comment by papb** [comment]
Hello @sean-mcmanus, sorry to take so long to reply. First of all I should note that I am just a beginner in C++, I'm trying to use VSCode because I'm already used to it (I'm a javascript developer). The odds are high that I'm the one making a mistake...

Instead of trying the v0.16.0-insiders2 that you mentioned, now that 8 days passed, I simply updated the extension to 0.16.1 instead. The problem persists, although I noticed something interesting, my `c_cpp_properties.json` file magically received two new rows, namely

```
"cStandard": "c11",
"cppStandard": "c++17"
```

but the problem persists. I also tried adding `"compilerPath": "C:/Cygwin/bin/g++"` / `"compilerPath": "C:/Cygwin/bin/gcc"` but the problem persists.

I also noticed that I might be doing something wrong by having `"intelliSenseMode": "clang-x64"` although I'm not using `clang` (I'm using `g++`), I don't even have a `clang.exe` in my `C:/Cygwin/bin` folder anyway. So I changed it to `"intelliSenseMode": "msvc-x64"` (although I admit I have no idea what I'm doing, just trying things around), and curiously, the **error changed!** Now, instead of squiggles below `std::`, I have this:

![https://i.imgur.com/npXM1ni.jpg](https://i.imgur.com/npXM1ni.jpg)

Note: I also tried changing `"cppStandard": "c++17"` to `c++14`, `c++11`, and even `c++03` but the error persists.

**Comment by sean-mcmanus** [comment]
clang-x64 is the expected mode for Cygwin...we have plans to add a gcc mode to make this less confusing (msvc-x64 is for cl.exe or clang on Windows).

If you temporarily set the loggingLevel to the hidden value of "6" and then open a new C++ file, you should be able to see the results of our compilerPath querying and what we send to our IntelliSense process in regards to includePath and defines.

I'll see if we can repro the issue ourselves with Cygwin.

**Comment by apo-egix** [comment]
i got a similar problem with mingw64

![prclyqq](https://user-images.githubusercontent.com/16270939/38271035-82bf5756-3785-11e8-8745-f1ac9cfbfe28.png)
![rsddamb](https://user-images.githubusercontent.com/16270939/38271036-830b589a-3785-11e8-818f-6d164b9cdfcc.png)

Scene is a custom class.
s_scenes is an instance of std::map.

g++ compiles it without any warning.

**Comment by sean-mcmanus** [comment]
@papb Good new -- I'm able to repro the bug. I'm able to fix it via adding #include <stddef.h> to the top of the file. I haven't figured out yet why that matters, but I believe it's because required compiler defines are missing (which are added in stddef.h when not passed in via the compiler), and the Cygwin gcc compiler doesn't accept "nul" as input, so the compilerPath feature fails to work for it...not sure what the alternative is yet.

@apo-egix I haven't looked at the mingw64 issue yet.

Bob and I are OOF till Monday -- maybe @grdowns  can answer questions until then.

**Comment by apo-egix** [comment]
@sean-mcmanus if i add `#include <stddef.h>` before `#include <map>` it resolves my issue aswell

**Comment by bobbrow** [comment]
You could also add `c:/path/to/stddef.h` to the `forcedInclude` array in **c_cpp_properties.json** so that you don't have to edit your source files.

**Comment by sean-mcmanus** [comment]
@papb We figured out how to get compilerPath to work with Cygwin, so we expect it to be available in our next release (it's very similar to the WSL case in that the Cygwin bash.exe needs to be called first and the "/usr" output converted to Windows paths). UPDATE: I got it working using `C:\\Cygwin64\\bin\\clang-5.0.exe` and `C:\\Cygwin64\\bin\\gcc.exe` without needing to call into bash...expected to be in our Insiders next week.

@apo-egix Your problem should be fixable via setting the compilerPath to the MinGW compiler -- that compiler should accept "nul". If this isn't working for then we could have another bug.

**Comment by apo-egix** [comment]
@bobbrow @sean-mcmanus both ways solved the issue
thanks guys

**Comment by bobbrow** [comment]
We have a fix for this in the [insiders build](https://github.com/Microsoft/vscode-cpptools/releases/tag/v0.17.0-insiders) if you would like to try it out and provide feedback.

**Comment by papb** [comment]
Cool, I will see if I can take a look later. Thanks!

**Comment by bobbrow** [comment]
This is fixed in 0.17.0.

**Comment by papb** [comment]
Indeed, it works :) Thank you all very much. Keep up the good work! :grin:
